File: benchmarks/python/scenarios/value_array_c_like.py
"""Pinned M3 value-array kernel and Python-semantics slow-path oracle."""

from typing import Any

import pcc


@pcc.valueclass
class Sample:
    x: float
    y: float


def make_samples() -> pcc.array[Sample, 2]:
    return pcc.array[Sample, 2](Sample(0.125, 0.5), Sample(0.25, 0.75))


def hot(values: pcc.array[Sample, 2], rounds: int) -> float:
    total: float = 0.25
    i: int = 0
    while i < rounds:
        left = values[0]
        right = values[1]
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        total = (total + left.x) * right.y - left.y
        total = (total + right.x) * left.y - right.y
        i = i + 1
    return total


def checked(values: pcc.array[Sample, 2], index: int) -> Sample:
    return values[index]


def escape(values: pcc.array[Sample, 2], index: int) -> Any:
    return values[index]


def main() -> None:
    print(hot(make_samples(), 1_000_000))  # PCC_M3_BENCHMARK_ROUNDS
    print(checked(make_samples(), -1).x)
    try:
        checked(make_samples(), 2)
    except IndexError:
        print("index-error")
    try:
        checked(make_samples(), 1 << 100)
    except OverflowError:
        print("overflow-error")
    boxed = escape(make_samples(), 1)
    alias = boxed
    print(boxed is alias)


main()

File: benchmarks/c/value_array_c_like.c
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>

typedef struct {
    double x;
    double y;
} Sample;

static double hot(const Sample values[static 2], int64_t rounds) {
    double total = 0.25;
    for (int64_t i = 0; i < rounds; i++) {
        const Sample left = values[0];
        const Sample right = values[1];
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
        total = (total + left.x) * right.y - left.y;
        total = (total + right.x) * left.y - right.y;
    }
    return total;
}

int main(void) {
    const Sample values[2] = {{0.125, 0.5}, {0.25, 0.75}};
    printf("%.17g\n", hot(values, INT64_C(1000000)));
    return 0;
}
